File: backend/src/routes/batismos.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const { getDb } = require('../db/schema');
const { autenticar, filtrarCongregacao } = require('../middleware/auth');
const { enfileirar } = require('../services/sync');

const router = express.Router();
router.use(autenticar, filtrarCongregacao);

router.get('/', (req, res) => {
  const db = getDb();
  const where = req.congregacaoFiltro ? `AND b.congregacao_id = '${req.congregacaoFiltro}'` : '';
  const rows = db.prepare(`
    SELECT b.*, m.nome as membro_nome, m.foto_url,
           p.nome as pastor_nome
    FROM batismos b
    JOIN membros m ON b.membro_id = m.id
    LEFT JOIN membros p ON b.pastor_id = p.id
    WHERE 1=1 ${where}
    ORDER BY b.data DESC
  `).all();
  res.json(rows);
});

// Membros não batizados
router.get('/pendentes', (req, res) => {
  const db = getDb();
  const where = req.congregacaoFiltro ? `AND congregacao_id = '${req.congregacaoFiltro}'` : '';
  const rows = db.prepare(`
    SELECT id, nome, foto_url, data_conversao, congregacao_id
    FROM membros
    WHERE status = 'ativo'
    AND (data_batismo IS NULL OR data_batismo = '')
    ${where}
    ORDER BY nome
  `).all();
  res.json(rows);
});

router.post('/', (req, res) => {
  const db = getDb();
  const { membro_id, data, local, pastor_id } = req.body;
  const congregacao_id = req.congregacaoFiltro || req.body.congregacao_id;
  const id = uuidv4();
  const agora = new Date().toISOString();

  db.prepare(`INSERT INTO batismos (id, congregacao_id, membro_id, data, local, pastor_id, criado_em)
    VALUES (?, ?, ?, ?, ?, ?, ?)`).run(id, congregacao_id, membro_id, data, local, pastor_id || null, agora);

  // Atualiza data_batismo no membro
  db.prepare('UPDATE membros SET data_batismo = ? WHERE id = ?').run(data, membro_id);

  const batismo = db.prepare(`
    SELECT b.*, m.nome as membro_nome, m.foto_url
    FROM batismos b JOIN membros m ON b.membro_id = m.id
    WHERE b.id = ?
  `).get(id);

  enfileirar('batismos', 'upsert', { id, congregacao_id, membro_id, data, local, pastor_id, criado_em: agora });
  res.status(201).json(batismo);
});

router.delete('/:id', (req, res) => {
  const db = getDb();
  const b = db.prepare('SELECT * FROM batismos WHERE id = ?').get(req.params.id);
  if (!b) return res.status(404).json({ erro: 'Não encontrado' });
  db.prepare('DELETE FROM batismos WHERE id = ?').run(req.params.id);
  // Remove data_batismo do membro
  db.prepare('UPDATE membros SET data_batismo = NULL WHERE id = ?').run(b.membro_id);
  res.json({ ok: true });
});

module.exports = router;

File: backend/app/routes/batismos.py
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import aliased, Session
from app.core.database import get_db
from app.deps import congregacao_filter, get_current_user
from app.models import Batismo, Membro, Usuario
from app.utils import new_id, parse_date

router = APIRouter(prefix="/batismos", tags=["batismos"])

class BatismoIn(BaseModel):
    membro_id: str
    data: str
    local: Optional[str] = None
    pastor_id: Optional[str] = None
    congregacao_id: Optional[str] = None

def _serializar(b: Batismo, membro: Membro, pastor: Optional[Membro]) -> dict:
    return {
        "id": b.id, "membro_id": b.membro_id, "congregacao_id": b.congregacao_id, "data": b.data,
        "local": b.local, "pastor_id": b.pastor_id, "membro_nome": membro.nome, "foto_url": membro.foto_url,
        "pastor_nome": pastor.nome if pastor else None,
    }

@router.get("")
def listar(db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user),
           cong_filtro: Optional[str] = Depends(congregacao_filter)):
    Pastor = aliased(Membro)
    q = (
        db.query(Batismo, Membro, Pastor)
        .join(Membro, Batismo.membro_id == Membro.id)
        .outerjoin(Pastor, Batismo.pastor_id == Pastor.id)
        .filter(Batismo.tenant_id == cu.tenant_id)
    )
    if cong_filtro:
        q = q.filter(Batismo.congregacao_id == cong_filtro)
    rows = q.order_by(Batismo.data.desc()).all()
    return [_serializar(b, m, p) for b, m, p in rows]

@router.get("/pendentes")
def pendentes(db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user),
              cong_filtro: Optional[str] = Depends(congregacao_filter)):
    q = db.query(Membro).filter(Membro.tenant_id == cu.tenant_id, Membro.status == "ativo",
                                 Membro.data_batismo.is_(None))
    if cong_filtro:
        q = q.filter(Membro.congregacao_id == cong_filtro)
    membros = q.order_by(Membro.nome).all()
    return [{"id": m.id, "nome": m.nome, "foto_url": m.foto_url, "data_conversao": m.data_conversao,
             "congregacao_id": m.congregacao_id} for m in membros]

@router.post("", status_code=201)
def criar(payload: BatismoIn, db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user),
          cong_filtro: Optional[str] = Depends(congregacao_filter)):
    membro = db.query(Membro).filter(Membro.id == payload.membro_id, Membro.tenant_id == cu.tenant_id).first()
    if not membro:
        raise HTTPException(status_code=404, detail="Membro não encontrado")
    cong_id = cong_filtro or payload.congregacao_id or membro.congregacao_id

    data_batismo = parse_date(payload.data)
    batismo = Batismo(id=new_id(), tenant_id=cu.tenant_id, congregacao_id=cong_id, membro_id=payload.membro_id,
                       data=data_batismo, local=payload.local, pastor_id=payload.pastor_id)
    db.add(batismo)
    membro.data_batismo = data_batismo
    db.commit(); db.refresh(batismo)

    pastor = db.query(Membro).filter(Membro.id == payload.pastor_id).first() if payload.pastor_id else None
    return _serializar(batismo, membro, pastor)

@router.delete("/{batismo_id}")
def remover(batismo_id: str, db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user)):
    b = db.query(Batismo).filter(Batismo.id == batismo_id, Batismo.tenant_id == cu.tenant_id).first()
    if not b:
        raise HTTPException(status_code=404, detail="Não encontrado")
    membro = db.query(Membro).filter(Membro.id == b.membro_id).first()
    if membro:
        membro.data_batismo = None
    db.delete(b); db.commit()
    return {"ok": True}
